### PR TITLE
[kimchi] absorb verifier index + recursion data

### PR DESCRIPTION
**Pull request by mimoo**
- [x] adds number of proofs verifier in a recursion circuit to `ConstraintSystem`
- [x] adds public input size to verifier index
- [x] absorbs verifier index in proof
- [x] absorbs prev challenges in proof
- [x] specify

closes https://github.com/o1-labs/proof-systems/issues/466 https://github.com/o1-labs/proof-systems/issues/552

### CODE REVIEW

File: kimchi/src/prover.rs [hunk 1]
```diff
@@ -191,6 +191,15 @@ where
         //~ 1. Setup the Fq-Sponge.
         let mut fq_sponge = EFqSponge::new(index.fq_sponge_params.clone());
 
+        //~ 1. Absorb the verifier index
+        let verifier_index = index.verifier_index();
```
Comment (mrmr1993) [inline kimchi/src/prover.rs:195]: Doesn't need to be done now, but we should compute this once while we're creating the index and cache the initialised sponge in the index. We don't really want to recommit to everything in the verifier index in every proof.

File: kimchi/src/verifier.rs [hunk 2]
```diff
@@ -63,6 +63,7 @@ impl<G: CommitmentCurve> ProverProof<G>
 where
     G::BaseField: PrimeField,
 {
+    /// ?
```
Comment (mrmr1993) [inline kimchi/src/verifier.rs:66]: Stop doing this.

File: kimchi/src/verifier_index.rs [hunk 3]
```diff
@@ -248,6 +258,58 @@ impl<G: CommitmentCurve> VerifierIndex<G>
...
+            for commitment in commitments {
+                sponge.absorb_g(&commitment.unshifted);
+            }
+
+            // TODO: should we absorb:
+            // - lookup_used
+            // - max_joint_size
+        }
+
+        // TODO: should we absorb:
```
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:306]: No

File: kimchi/src/verifier_index.rs [hunk 4]
```diff
@@ -248,6 +258,58 @@ impl<G: CommitmentCurve> VerifierIndex<G>
...
+                .chain(lcs.lookup_selectors.lookup_gate.iter())
+                .chain(lcs.lookup_selectors.range_check_gate.iter())
+                .chain(lcs.table_ids.iter())
+                .chain(lcs.runtime_tables_selector.iter());
+
+            for commitment in commitments {
+                sponge.absorb_g(&commitment.unshifted);
+            }
+
+            // TODO: should we absorb:
```
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:301]: No

File: kimchi/src/verifier_index.rs [hunk 5]
```diff
@@ -248,6 +258,58 @@ impl<G: CommitmentCurve> VerifierIndex<G>
...
+            .chain([&self.psm_comm])
+            .chain([&self.complete_add_comm])
+            .chain([&self.mul_comm])
+            .chain([&self.emul_comm])
+            .chain([&self.endomul_scalar_comm])
+            .chain(self.chacha_comm.iter().flatten())
+            .chain(&self.range_check_comm);
+
+        for commitment in commitments {
+            sponge.absorb_g(&commitment.unshifted);
```
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:282]: I think (for pickles reasons) we want to hash the verifier index and then absorb that hash. Can you change this to do that?

File: kimchi/src/verifier_index.rs [hunk 6]
```diff
@@ -183,6 +190,9 @@ where
                 cell
             },
 
+            public_input_size: self.cs.public,
+            recursion_challenges: todo!(),
```
Comment (mrmr1993) [inline kimchi/src/verifier_index.rs:194]: Take it as an input to the constraint system builder, pass it through from there.

**Review by mrmr1993 — CHANGES_REQUESTED** [review]

File: kimchi/src/verifier.rs (hunk 2 above)
Comment (mimoo) [reply kimchi/src/verifier.rs:66]: it's a WIP PR, I can do whatever I want :P

Comment (mrmr1993) [reply kimchi/src/verifier.rs:66]: You did ask me to look at it!

**Comment by mimoo** [comment]
Interestingly, I expected being able to create a proof that has prev_challenges of random length (at least as long as it's smaller than our SRS size, why would there be an issue?) 
But tests don't pass when I try that. From a quick glance I think the verifier's code can handle prev_challenges from a higher domain, but not from a smaller domain:  https://github.com/o1-labs/proof-systems/blob/e123cf8362c57191c4ab96501eb80a069e3a8aaa/kimchi/src/verifier.rs#L66

investigating...

**Pull request closed without merging**